File: app/models/routine.rb
class Routine < ApplicationRecord
  include Discard::Model

  belongs_to :supervisor, class_name: 'User', inverse_of: :supervised_routines, required: false
  belongs_to :patient, class_name: 'User', inverse_of: :routines, required: false

  has_many :routine_exercises
  has_many :routine_intents
  has_many :exercises, through: :routine_exercises, inverse_of: :routines

  accepts_nested_attributes_for :routine_exercises, allow_destroy: true, reject_if: :all_blank

  scope :supervised_by, -> (user) {
    where(supervisor: user)
  }

  scope :owned_by, -> (user) {
    where(patient: user)
  }
end

File: app/javascript/resources/Routine.js
import APILibrary from "@/resources/APILibrary";
import BaseResource from "@/resources/BaseResource";

import "./Exercise";
import "./RoutineExercise";

class Routine extends BaseResource {
  // required by active-resource
  // https://github.com/nicklandgrebe/active-resource.js/issues/47
  static className = "Routine";
  static queryName = "routines";

  // required by @casl/ability
  // https://stackoverflow.com/questions/57677510/casl-is-not-working-properly-in-vue-production-mode
  static modelName = "Routine";

  static define() {
    this.attributes("id", "supervisorId", "patientId");

    // this.belongsTo("patient", { className: "User" });
    // this.belongsTo("supervisor", { className: "User" });

    this.hasMany("exercises");
    this.hasMany("routineExercises");
  }
}

export default APILibrary.createResource(Routine);
